File: p454.cpp
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    int fourSumCount(vector<int>& A, vector<int>& B, vector<int>& C, vector<int>& D) {
        int counter = 0;
        int n = A.size();
        int m = B.size();
        int o = C.size();
        int p = D.size();
        map<int, int> counts;
        for(int i = 0; i < n; ++i) {
            for(int j = 0; j < m; ++j) {
                ++counts[A[i]+B[j]];
            }
        }
        for(int k = 0; k < o; ++k) {
            for(int l = 0; l < p; ++l) {
                counter += counts[-(C[k]+D[l])];
            }
        }
        return counter;
    }
};

File: p454.py
from collections import Counter
from typing import List

class Solution:
    def fourSumCount(self, A: List[int], B: List[int], C: List[int], D: List[int]) -> int:
        c = Counter()
        ans = 0

        for num1 in A:
            for num2 in B:
                c[num1+num2] += 1

        for num1 in C:
            for num2 in D:
                if -(num1+num2) in c:
                    ans += c[-(num1+num2)]

        return ans
